File: angel-ps/python/pyangel/ml_runner.py
#
# Tencent is pleased to support the open source community by making Angel available.
#
# Copyright (C) 2017-2018 THL A29 Limited, a Tencent company. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in
# compliance with the License. You may obtain a copy of the License at
#
# https://opensource.org/licenses/Apache-2.0
#
# Unless required by applicable law or agreed to in writing, software distributed under the License is
# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
# either express or implied. See the License for the specific language governing permissions and
#

import os
from abc import ABCMeta, abstractmethod

from pyangel.conf import AngelConf
from pyangel.ml.client.angel_client_factory import AngelClientFactory

class MLRunner(object):
    """
    Training job to obtain a model
    """

    def train(self, conf, model, class_name):
        """
        Default train method with standard training process. Don't Override this method
        :param conf: Configuration for angel application
        :param model: Data training model
        :param train_class: train task class name
        :return:
        """
        # Get Java HashMap instance which converted from a python dict
        jmap = conf.dict_to_jmap()
        client = AngelClientFactory.get(jmap, conf)

        client.startPSServer()
        client.loadModel(model)
        client.runTask(class_name)
        client.waitForCompletion()
        client.saveModel(model)

        client.stop()

    @abstractmethod
    def inc_train(self, conf):
        """
        Incremental training job to obtain a model based on a trained model
        :param conf:
        :return:
        """
        pass

    def predict(self, dconf, model, class_name):
        """
        Default predict method with standard predcit process. Don't try to override this Method
        :param conf:
        :return:
        """
        jmap = Configuration._dict_to_jmap(dconf)
        client = AngelClientFactory.get(jmap)
        client.startPSServer()
        client.loadModel(model)
        client.runTask(class_name)
        client.waitForCompletion()

        client.stop()

    def confirm_path(self, file_path, delete):
        """
        Judge the file path is right or doesn't exists
        :param file_path:
        :return:
        """
        if os.path.exists(file_path):
            if delete:
                rmr = raw_input("{} is already exists, do you want to remove it?[y/n]: ".format(file_path))
                if rmr == "y":
                    os.remove(file_path)
        else:
            print("Directory {} will be created".format(file_path))

File: angel-ps/mllib/src/main/scala/com/tencent/angel/ml/core/MLRunner.scala
/*
 * Tencent is pleased to support the open source community by making Angel available.
 *
 * Copyright (C) 2017-2018 THL A29 Limited, a Tencent company. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in 
 * compliance with the License. You may obtain a copy of the License at
 *
 * https://opensource.org/licenses/Apache-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 *
 */


package com.tencent.angel.ml.core

import com.tencent.angel.AppSubmitter
import com.tencent.angel.client.AngelClientFactory
import com.tencent.angel.conf.AngelConf
import com.tencent.angel.ml.core.conf.MLConf
import com.tencent.angel.ml.model.MLModel
import com.tencent.angel.worker.task.BaseTask
import org.apache.hadoop.conf.Configuration


trait MLRunner extends AppSubmitter {

  /**
    * Training job to obtain a model
    */
  def train(conf: Configuration)

  /**
    * Default train method with standard training process. Don't Override this method.
    *
    * @param conf
    * @param model
    * @param taskClass
    */
  final protected def train(conf: Configuration, model: MLModel, taskClass: Class[_ <: BaseTask[_, _, _]]): Unit = {
    val client = AngelClientFactory.get(conf)

    try {
      client.startPSServer()
      client.loadModel(model)
      client.runTask(taskClass)
      client.waitForCompletion()
      client.saveModel(model)
    } finally {
      client.stop()
    }
  }

  /**
    * Using a model to predict with unobserved samples
    */
  def predict(conf: Configuration)

  /**
    * Default predict method with standard predict process. Don't try to override this Method.
    *
    * @param conf
    * @param model
    * @param taskClass
    */
  final protected def predict(conf: Configuration, model: MLModel, taskClass: Class[_ <: BaseTask[_, _, _]]): Unit = {
    val client = AngelClientFactory.get(conf)
    try {
      client.startPSServer()
      client.loadModel(model)
      client.runTask(taskClass)
      client.waitForCompletion()
    } finally {
      client.stop(0)
    }
  }

  @throws[Exception]
  override
  def submit(conf: Configuration): Unit = {
    val actType = conf.get(AngelConf.ANGEL_ACTION_TYPE)
    actType match {
      case MLConf.ANGEL_ML_TRAIN | MLConf.ANGEL_ML_INC_TRAIN  =>
        train(conf)
      case MLConf.ANGEL_ML_PREDICT =>
        predict(conf)
      case _ =>
        println("Error action type, should be train or predict or inctrain")
        System.exit(1)
    }
  }

}
